### ISSUE TITLE
avoid code duplication in persistence impl.

### ISSUE THREAD

**Issue by pine3ree** [mezzio/mezzio-session#8]
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes (refactoring)
| RFC         | no
| BC Break    | no

#### Summary

Hello, having implemented the cache-limiter generation feature in mezzio-session-ext, I found out that the same code is duplicated in mezzio-session-cache.
This is because php-session is one extension that generates a response cookie and cache-header always in the same way (with custom params)  and allowing us to attach different save-handlers, while mezzio-session persistence layer is more generic and takes care both of the persistence of the session-data and the session-identification in the client-browser. mezzio-session does not assume that phpsession-like headers must be used in all persistence implementations. 

Some of these implementations may have common codebase.
IMO it would be better to have either traits for such common features or cache-headers-generator/set-cookie-header-generator services as dependencies for the persistence implementations that require them.

kind regards

**Comment by weierophinney** [comment]
Considering both of those implementations depend on this library, please feel free to extract the code to a trait or traits here. We can then gradually update the minimum requirement in each of those so we can make use of the trait(s).

**Comment by pine3ree** [comment]
Hello @weierophinney ,

I was thinking about creating a couple of traits in this package mezzio-session (or possibly later moved in their own package mezzio-session-persistence):

`Mezzio\Session\Persistence\CacheHeadersGeneratorTrait`
`Mezzio\Session\Persistence\SessionCookieAwareTrait` (generic implementation for reading the session id cookie from the request and set the session cookie header into the response)

what changes in the the implementation is where we take the default parameters (e.g ini_get for mezzio-session-ext).

**Comment by weierophinney** [comment]
@pine3ree Go for it! Ping me for review when you have a PR ready.

**Comment by pine3ree** [comment]
Hello @weierophinney,

I have added the initial commit for trait-extraction:

https://github.com/pine3ree/mezzio-session/tree/add-persistence-traits/src/Persistence

no tests so far as there are a few points I'd wish to discuss before an actual PR.


#### 1. Nomenclature

I want the methods used in consumer classes to be as self-explanatory and complete as possible:

`SessionCookieAwareTrait::getSessionCookieFromRequest()` would replace `CacheSessionPersistence::getCookieFromRequest()` and be used in `PhpSessionPersistence` class as well.

`SessionCookieAwareTrait::addSessionCookieToResponse()` would replace `PhpSessionPersistence::addSessionCookie()` and be used in `CacheSessionPersistence` class as well.

`CacheHeadersGeneratorTrait::addCacheHeadersToResponse()` would replace `PhpSessionPersistence::addCacheHeaders()` and be used in `CacheSessionPersistence` class as well.

I am not sure which is best here:
- `add(SessionCookie|CacheHeaders)ToResponse`
- `set(SessionCookie|CacheHeaders)IntoResponse` 
- `getResponseDecoratedWith(SessionCookie|CacheHeaders)` since it returns a response

maybe it's better that you decide, English is not my first lang.

#### 2. Previous class constants

Since we are using traits we have 2 choices:
  - use private static props and @internal public accessors for tests (this is what you find in the CacheHeadersGeneratorTrait)
  - use an extra helper class (@[see](https://github.com/pine3ree/mezzio-session/tree/add-persistence-traits/src/Persistence) the [`Mezzio\Persistence\Http`](https://github.com/pine3ree/mezzio-session/tree/add-persistence-traits/src/Persistence) example class) whose constants can be consumed in every class.
  
#### 3. SessionCookieAwareTrait::getSessionCookieLifetime()

This method would replace both: `PhpSessionPersistence::getCookieLifetime()` and `CacheSessionPersistence::getPersistenceDuration()` , we just need to initialize the prop `$this->cookieLifetime` in the constructor:

```php
// CacheSessionPersistence
$this->cookieLifetime = $this->persistent ? $this->cacheExpire : 0;
```
```php
/// PhpSessionPersistence
$this->cookieLifetime = (int) ini_get('session.cookie_lifetime');
```
All other props are initialized on construction either by config params and/or by `ini_get()` calls or other means in other implementations using the `SessionCookieAwareTrait`

#### 4. CacheHeadersGeneratorTrait::getLastModified()

This (first defined in `PhpSessionPersistence`) is now caching the computed value at first request, same way it was done in `CacheSessionPersistence::determineLastModifiedValue()`, method that it would replace. I kept my original implementation for a reason: `determineLastModifiedValue` is assuming hardcoded values `'public/index.php'`, `'index.php'` defaulting to filemtime the cwd, while `getlastmod()`  always checks the currently running script, so if I want this to be `'public/app.php'` or `'public/swoole.php'` it would target those ones (or `filemtime(__FILE__)` as a default in the method). I would like that future versions of laminas and mezzio packages didn't have hardcoded paths. So far I have had troubles changing the `config` dir to `etc` (development-tools) and the `public` dir to `web` (this last issue with mezzio-swoole)...but hey, I am a big fan of linux (3-letters) FHS.

#### 5. PHP-version requirements
mezzio-session requires php^7.1
mezzio-session-ext requires php^7.3 and dflydev/fig-cookies:^2.0
mezzio-session-cache requires php-7.1 and dflydev/fig-cookies:^1.0 ||^2.0

with these traits (if not moved into different package) we would have the same dflydev/fig-cookies requirement of mezzio-session-cache , so I added a check (https://github.com/pine3ree/mezzio-session/blob/add-persistence-traits/src/Persistence/SessionCookieAwareTrait.php#L110) for same-site capability when generating the response SetCookie. Not sure it's that elegant...and one of those two conditions would be enough.


##### X. Unrelated

I was browsing this code:

https://github.com/mezzio/mezzio-session-cache/blob/master/src/CacheSessionPersistence.php#L189

Why is the session-id regenerated even if only the session-data is changed? Shouldn't we expect the cache adapter to just overwrite the data stored under the old key (session-id)? I have not used the CacheSessionPersistence yet, so maybe it's a silly question and I am just missing a trivial point.

kind regards

**Comment by weierophinney** [comment]
**1. Nomenclature**: Your suggestions all make sense to me!

**2. Previous class constants**: I like the `Mezzio\Persistence\Http` approach.

**3. SessionCookieAwareTrait::getSessionCookieLifetime()**: Sounds good.

**4. CacheHeadersGeneratorTrait::getLastModified()**: Honestly, this was a difficult context when I was doing the original development. I'm not terribly happy with the current filesystem-based solution, but I ran out of ideas. Use your best judgment; the goal is to ensure there is a sane default that we can communicate to users easily.

** 5. PHP-version requirements**: Since this would be a new minor version, we can bump the minimum supported PHP version easily (per our decisions at the first TSC meeting last month). The main thing to remember is that when we bump, we have to support that minor tree until the minimum supported PHP version is no longer supported.

** X. Unrelated*: It's to prevent a session-fixation vector.

**Comment by pine3ree** [comment]
Hello @weierophinney, I was almost sure I forgot one last point (but it was 5 a.m. here).

#### 2 Mezzio\Persistence\Http
Good, this is actually what I did in the first place...

#### 6 Cache-limiters

In `CacheHeadersGeneratorTrait::generateCacheHeaders()` I kept my original code allowing a not recognized cacheLimiter value and not sending any cache-limiter headers, because this is also the behaviour of php ext-session. Setting for instance ini_set('session.cache_limiter', '') before starting the session has the effect of not sending any cache-limiter header. Of course different implementations may initialize a default value. 

##### 4 Filesystem structure
This is just an idea, maybe will discuss it later in the proper repo...
1. we could add the FS layout in the configuration using `__DIR__`  in the main config file to set the config-directoy value
2. Define and read the filesystem structure in composer.json file. The app-skeleton installer and the component-installer would then use these entries to set for instance the dir-layout config values in an auto-generated `config/autoload/dir-layout.local.php`... 

##### X. Unrelated:
That is of course the best security option but I have 2 doubts:
1. Shouldn't we decide programmatically when to regenerate the id like for other implementations?
   In mezzion-session-ext we use this condition:
```php
if ($session->isRegenerated() || ($id === '' && $session->hasChanged()) {...}
```
2. Writing to a different cache entry every time the session data has been changed may result in a lot of stale entries (I am thinking about a very busy app) unless we also remove the `replaced` one  in `persistSessionDataToCache` passing into it the old id.

Thank You. Kind regards.

**Comment by weierophinney** [comment]
**6. Cache-limiters**: Makes sense to me.

**4. Filesystem structure**: The problem with option (1) is it depends on the user setting a config value **accurately**. We'd definitely still need a sane fallback. With regards to (2) if you're going to look at `composer.json`... why not just use _that_ file? :smile: 

**X. Unrelated**:

1. Programmatically is fine. The main point is, if a change is detected, regenerate the ID.

2. I brought that up with some people while I was writing the library, and they indicated that most caching systems will expire entries not accessed after X amount of seconds, making it a mostly non-issue. One thing to note: if we _were_ to manually expire it/remove it, we add immediate performance overhead, which is another argument against doing it then. We _could_ provide a CLI tool that could be run periodically to expire session cache items, however, which would give users a mechanism to do it on their own terms.

**Comment by pine3ree** [comment]
Hello @weierophinney,

I made a few minor code and nomenclature changes so I think the `src` code can be considered ready or close to ready. I'll ping @michalbundyra for the preferred nomenclature as discussed in the first comments.

About `test` cases:

since we provide traits with private methods only we should rely on the consumer classes tests, but I would prefer to add testing cases as well. Would it be acceptable to construct consumer classes like this:

```php
class CacheHeadersGeneratorTraitTest extends TestCase
{
    //...

    protected function createConsumerInstance(
        int $cacheExpire = 180,
        string $cacheLimiter = '',
        string $lastModified = null
    ) {
        return new class($cacheExpire, $cacheLimiter, $lastModified) {

            use CacheHeadersGeneratorTrait {
                addCacheHeadersToResponse as trait_addCacheHeadersToResponse;
                responseAlreadyHasCacheHeaders as trait_responseAlreadyHasCacheHeaders;
                getLastModified as trait_getLastModified;
            }

            public function __construct(int $cacheExpire, string $cacheLimiter, string $lastModified = null)
            {
                $this->cacheExpire  = $cacheExpire;
                $this->cacheLimiter = $cacheLimiter;
                $this->lastModified = $lastModified;
            }

            public function addCacheHeadersToResponse(ResponseInterface $response) : ResponseInterface
            {
                return $this->trait_addCacheHeadersToResponse($response);
            }

            public function getLastModified()
            {
                return $this->trait_getLastModified();
            }

            public function responseAlreadyHasCacheHeaders(ResponseInterface $response) : bool
            {
                return $this->trait_responseAlreadyHasCacheHeaders($response);
            }
        };
    }
    //...
}
```

with proxy methods accessing the trait methods?
kind regards,
maks

**Comment by pine3ree** [comment]
update:
I also have the updates for of `mezzio-session-ext` (src+test) using the new persistence traits
tonight i will start doing the same for `mezzio-session-cache`

**Comment by pine3ree** [comment]
Hello @weierophinney, I completed  (locally) the update of `mezzio-session-cache` to use the new persistence-traits. I also added a new repo (https://github.com/pine3ree/mezzio-session/tree/add-persistence-traits-tests) with sample test cases (introduced in https://github.com/mezzio/mezzio-session/issues/8#issuecomment-610730866), pls check if they make sense to you.

PS
I also added the `Max-Age` token in the response session cookie header when the `Expires` token is present, since this is also php-session's behavior.

kind regards
